### PR TITLE
Add Support for `scroll_event` in WebAgg and NbAgg [backport to 1.4.x]

### PR DESCRIPTION
**Pull request by blink1073**
Attached a scroll event handler to `canvas_div` and added the appropriate handling to `backend_webagg_core`.

### CODE REVIEW

**Comment by blink1073** [comment]
Wow, that felt like a battle with Travis with all the guess-and-checking.

File: lib/matplotlib/backends/web_backend/mpl.js [hunk 1]
```diff
@@ -402,7 +419,8 @@ mpl.figure.prototype.mouse_event = function(event, name) {
     var x = canvas_pos.x;
     var y = canvas_pos.y;
 
-    this.send_message(name, {x: x, y: y, button: event.button});
+    this.send_message(name, {x: x, y: y, button: event.button,
+                             step: event.step});
```
Comment (tacaswell) [inline lib/matplotlib/backends/web_backend/mpl.js]: what done js do if you have not set a value for `event.step`?  I assume magically pass a null-like value...

File: lib/matplotlib/backends/web_backend/mpl.js [hunk 2]
```diff
@@ -125,6 +125,23 @@ mpl.figure.prototype._init_canvas = function() {
...
+        } else if (event.detail) { /** Mozilla case. */
+            /** In Mozilla, sign of delta is different than in IE.
+             * Also, delta is multiple of 3.
+             */
+            event.step = -event.detail/3;
+        }
+        mouse_event_fn(event);
+    }
+
+    $(canvas_div)[0].addEventListener('DOMMouseScroll', scroll_fn, false);
```
Comment (tacaswell) [inline lib/matplotlib/backends/web_backend/mpl.js]: Can you just do `canvas_div.addEventListener(...)` here?  We have (what I think) is a reference to the js version of the div so there is no need to look it up.

Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: Simplified now, does not work without the `[0]` access though.

File: lib/matplotlib/backends/web_backend/mpl.js (hunk 1 above)
Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: JS assigned it to `undefined` which is translated to `None` AFAIK.

File: lib/matplotlib/backends/web_backend/mpl.js (hunk 2 above)
Comment (tacaswell) [inline lib/matplotlib/backends/web_backend/mpl.js]: Then why do the next two lines work?

Comment (tacaswell) [inline lib/matplotlib/backends/web_backend/mpl.js]: That wasn't meant to be hostile, I am in fact quite confused.

Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: I'm as confused as you are.

Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: All I know is that `canvas_div` has no attribute `addEventListener`, but `canvas_div[0]` does.

File: lib/matplotlib/backends/web_backend/mpl.js [hunk 3]
```diff
@@ -125,6 +125,23 @@ mpl.figure.prototype._init_canvas = function() {
...
+            /** In Mozilla, sign of delta is different than in IE.
+             * Also, delta is multiple of 3.
+             */
+            event.step = -event.detail/3;
+        }
+        mouse_event_fn(event);
+    }
+
+    canvas_div[0].addEventListener('DOMMouseScroll', scroll_fn, false);
+    canvas_div[0].addEventListener('mousewheel', scroll_fn, false);
```
Comment (tacaswell) [inline lib/matplotlib/backends/web_backend/mpl.js]: My current theory is that the other functions are jquery sourced methods (and `canvas_div` is a jquery object) where as `addEventListener` is a native javastcript function which needs to be called on the actual html element (?) which is the first object returned when you index into a jquery object.  I _think_ we should be using `canvas_div.scroll(scroll_fun)` (https://api.jquery.com/scroll/)  here (might have to fake up the `canval_div` to be 'scrollable' which hopefully we can do without making the browser try to scroll it.

File: lib/matplotlib/backends/backend_webagg_core.py [hunk 4]
```diff
@@ -190,7 +190,7 @@ def handle_event(self, event):
             pass
         elif e_type == 'draw':
             self.draw()
-        elif e_type in ('button_press', 'button_release', 'motion_notify'):
+        elif e_type in ('button_press', 'button_release', 'motion_notify', 'scroll'):
```
Comment (pelson) [inline lib/matplotlib/backends/backend_webagg_core.py]: Line too long is causing a test failure.

**Comment by pelson** [comment]
Nice. :+1:

File: lib/matplotlib/backends/web_backend/mpl.js (hunk 3 above)
Comment (mdboom) [inline lib/matplotlib/backends/web_backend/mpl.js]: Yes -- to confirm @tacaswell's last comment -- let's do it the "jquery way" whenever possible to be consistent with tht rest of the js.

Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: I tried the following combos:  `canvas_div.scroll`, `canvas_div.on("scroll",`, `canvas.scroll`, `canvas.on("scroll"`, and none of them work.

Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: That is, `scroll_fn` is never called (I added a `console.log` in there).

**Comment by tacaswell** [comment]
This now has conflicts with the enter/leave event PR.

**Comment by blink1073** [comment]
Fixed (I hope).

File: lib/matplotlib/backends/web_backend/mpl.js (hunk 3 above)
Comment (tacaswell) [inline lib/matplotlib/backends/web_backend/mpl.js]: It looks like jquery is being too smart for it's own good here?

@minrk any words of wisdom on making scroll events play nice?

Comment (tacaswell) [inline lib/matplotlib/backends/web_backend/mpl.js]: https://developer.mozilla.org/en-US/docs/Web/Events/mousewheel <- this seems rather down on 'mousewheel'.

`canvas_div.on('wheel mousewheel', scroll_fn)` causes the events to be triggered (I was wrong about what 'scroll' meant), but the events you get back are wrapped jquery events so of course none of the current code works correctly (it ends up passing `None` through for the position which causes the python to barf when you do math with it).

Comment (blink1073) [inline lib/matplotlib/backends/web_backend/mpl.js]: I *think* `mousewheel` is still required by some browsers.  I updated the handler and its a lot cleaner.  I think it should still be tested in IE.  I'll boot into Windows tonight and try it out.

**Comment by blink1073** [comment]
The current code works in Firefox, Chrome, and IE.  Turns out we did not need "mousewheel".

**Comment by tacaswell** [comment]
Do we have any brave mac users who want to help out with safari?

**Comment by tonysyu** [comment]
@blink1073: How are you testing this? (I'm not talking about real tests, just a manual verification that things are working)

**Comment by blink1073** [comment]
@tonysyu, start `ipython` and `%paste` this:

```python
import matplotlib
import itertools
import numpy as np
matplotlib.use('webagg')
import matplotlib.pyplot as plt
plt.close('all')
fig, ax = plt.subplots()
x = np.linspace(0,10,100)
y = np.sin(x)
ln, = ax.plot(x,y, 'o', picker=5)
evt = []
colors = iter(itertools.cycle(['r', 'g', 'b', 'k', 'c']))
def on_event(event):
    evt.append(event)
    ln.set_color(next(colors))
    fig.canvas.draw()
    fig.canvas.draw_idle()
fig.canvas.mpl_connect('scroll_event', on_event)
plt.show()
```

Then click on the figure and scroll and make sure it changes colors.

**Comment by tonysyu** [comment]
Does the suggestion to start `ipython` and paste actually work on your system? IPython should choose a backend at start-up so the `matplotlib.use` call shouldn't have an effect. At least it doesn't on my system.

In any case, when run outside of IPython, the script works for me on Safari and Chrome on OSX. :+1: 

It'd be nice to have an example that can be fired up from the examples directory that tests some of these events that are difficult to test properly. I'm not necessarily suggesting it for this PR, just throwing it out there.

**Comment by blink1073** [comment]
@tonysyu, thanks!  (I forgot to mention I had to comment out the matplotlib import in my ipython startup file).  I agree about the demos.

**Comment by tacaswell** [comment]
@honesty I think that the nbagg_uat.nb is that demo (which does not yet
have these tests in it)

**Comment by tacaswell** [comment]
@tonysyu darn you auto correct

**Comment by blink1073** [comment]
I was waiting for #3965 to merge before updating the UAT, now done.

**Comment by tacaswell** [comment]
back-ported as 22eb1d57b2d99a5d16ce930baa84cac932804e29

**Pull request merged**